### PR TITLE
Support for cnpm and use electron as devDependence instead of electro…

### PR DESCRIPTION
**Pull request by wenshiqi0**
I am using `cnpm` instead of `npm` or `yarn` to install my node modules, and I find out that electron-rebuild does not support for `cnpm`. `cnpm` use link to handle the dependencies, as I can see electron-rebuild may make circles while scanning the dependencies tree and will rebuild a module twice and this will make node-gyp throw errors.

### CODE REVIEW

**Comment by fengmk2** [comment]
pnpm is using link too https://github.com/pnpm/pnpm

File: src/rebuild.ts [hunk 1]
```diff
@@ -42,6 +44,8 @@ class Rebuilder {
     this.ABI = nodeAbi.getAbi(electronVersion, 'electron');
     this.prodDeps = extraModules.reduce((acc, x) => acc.add(x), new Set());
     this.rebuilds = [];
+    this.realModulePaths = [];
```
Comment (MarshallOfSound) [inline src/rebuild.ts]: For performance reasons these two path arrays should be sets

**Comment by fengmk2** [comment]
@zcbenz we find the root cause finally.

File: src/rebuild.ts [hunk 2]
```diff
@@ -184,19 +188,40 @@ class Rebuilder {
   }
 
   rebuildAllModulesIn(nodeModulesPath: string, prefix = '') {
-    d('scanning:', nodeModulesPath);
+    // While we use `cnpm`, it will make a circle scanning the dep tree.
+    // We also need to ensure that the `node_modules` which we are scanning has never came before.
+    const realNodeModulesPath = fs.realpathSync(nodeModulesPath);
+    if (this.realNodeModulesPaths.indexOf(realNodeModulesPath) > -1) {
+      return;
+    } else {
```
Comment (MarshallOfSound) [inline src/rebuild.ts]: No need for this else statement as their is a `return` in the main if condition

File: src/rebuild.ts [hunk 3]
```diff
@@ -184,19 +188,40 @@ class Rebuilder {
...
+    for (const modulePath of fs.readdirSync(realNodeModulesPath)) {
+      // If we use `cnpm` to install mudules, it will be rebuilded failed,
+      // because of some same modules rebuilded twice or more.
+      // Need to ensure one dep only be rebuilded once.
+      const finalPath = path.resolve(nodeModulesPath, modulePath);
+      const realPath = fs.realpathSync(finalPath);
+
+      if (this.realModulePaths.indexOf(realPath) > -1) {
+        continue;
+      } else {
```
Comment (MarshallOfSound) [inline src/rebuild.ts]: Again no need for the else block as their is a `continue` statement in the main if block

File: src/rebuild.ts [hunk 4]
```diff
@@ -42,6 +44,8 @@ class Rebuilder {
     this.ABI = nodeAbi.getAbi(electronVersion, 'electron');
     this.prodDeps = extraModules.reduce((acc, x) => acc.add(x), new Set());
     this.rebuilds = [];
+    this.realModulePaths = [];
+    this.realNodeModulesPaths = [];
```
Comment (fengmk2) [inline src/rebuild.ts]: use `Map` instead `Array` will be more faster.

Comment (MarshallOfSound) [inline src/rebuild.ts]: See my comment below about using sets, all we need is a concept of "has been used before", we don't need to store a value with the path so maps will be less efficient than using a raw set

**Comment by MarshallOfSound** [comment]
Also if you pull in latest master the travis builds should go green 👍

**Comment by wenshiqi0** [comment]
Need some Windows test.

**Comment by wenshiqi0** [comment]
Test passed, but I find some more problems. Did electron-rebuild will handle `fsevents` rebuilding for Windows? Event I am in Windows, the `fsevents` module may still be rebuilded, and this make errors. @MarshallOfSound

**Pull request merged**